crates/proto/src/quic/quic_socket.rs

```diff
@@ -27,51 +32,20 @@ impl<S: DnsUdpSocket + QuicLocalAddr> Debug for QuinnAsyncUdpSocketAdapter<S> {
 }
 
 /// TODO: Naive implementation. Look forward to future improvements.
-impl<S: DnsUdpSocket + QuicLocalAddr + 'static> AsyncUdpSocket for QuinnAsyncUdpSocketAdapter<S> {
-    fn poll_send(
-        &self,
-        _state: &quinn::udp::UdpState,
-        cx: &mut Context<'_>,
-        transmits: &[quinn::udp::Transmit],
-    ) -> Poll<std::io::Result<usize>> {
-        // logics from quinn-udp::fallback.rs
-        let io = &self.io;
-        let mut sent = 0;
-        for transmit in transmits {
-            match io.poll_send_to(cx, &transmit.contents, transmit.destination) {
-                Poll::Ready(ready) => match ready {
-                    Ok(_) => {
-                        sent += 1;
-                    }
-                    // We need to report that some packets were sent in this case, so we rely on
-                    // errors being either harmlessly transient (in the case of WouldBlock) or
-                    // recurring on the next call.
-                    Err(_) if sent != 0 => return Poll::Ready(Ok(sent)),
-                    Err(e) => {
-                        if e.kind() == std::io::ErrorKind::WouldBlock {
-                            return Poll::Ready(Err(e));
-                        }
+impl<S: DnsUdpSocket + QuicLocalAddr + Clone + Sync + 'static> AsyncUdpSocket
+    for QuinnAsyncUdpSocketAdapter<S>
+{
+    fn create_io_poller(self: Arc<Self>) -> Pin<Box<dyn UdpPoller>> {
+        Box::pin(UdpPollHelper::new(move || {
+            let socket = self.io.clone();
+            async move { socket.writable().await }
+        }))
+    }
 
-                        // Other errors are ignored, since they will ususally be handled
-                        // by higher level retransmits and timeouts.
-                        // - PermissionDenied errors have been observed due to iptable rules.
-                        //   Those are not fatal errors, since the
-                        //   configuration can be dynamically changed.
-                        // - Destination unreachable errors have been observed for other
-                        // log_sendmsg_error(&mut self.last_send_error, e, transmit);
-                        sent += 1;
-                    }
-                },
-                Poll::Pending => {
-                    return if sent == 0 {
-                        Poll::Pending
-                    } else {
-                        Poll::Ready(Ok(sent))
-                    }
-                }
-            }
-        }
-        Poll::Ready(Ok(sent))
+    fn try_send(&self, transmit: &quinn::udp::Transmit<'_>) -> io::Result<()> {
+        self.io.try_io(Interest::WRITABLE, || {
+            self.inner.send((&self.io).into(), transmit)
+        })
```

Comment:
The prev version had a comment:
> // logics from quinn-udp::fallback.rs

However, the current version of [`quin-udp/src/fallback.rs`](https://github.com/quinn-rs/quinn/blob/2ad6dbbdcf9455c1165af6335803bab89fbb7f32/quinn-udp/src/fallback.rs) upstream doesn't hold much resemblance to the old, or updated, code in this PR.

Instead it seems like the new code is being cribbed from [`quinn/src/runtime/tokio.rs`](https://github.com/quinn-rs/quinn/blob/2ad6dbbdcf9455c1165af6335803bab89fbb7f32/quinn/src/runtime/tokio.rs). In that instance `self` is a quinn `UdpSocket` and `inner` is a quinn `udp::UdpSocketState`.

Can you speak more about the direction you were heading here? This seems to be where the majority of the build issues are (`self.io_try_io` doesn't exist, there's no `self.inner`, an there's a future that can't be shared between threads, etc). I'm not familiar enough with trust-dns, quinn, or async Rust to intuit the next step on my own. Is there a particular Quinn commit/changeset this is adapting to that I could look at?

Reply:
I haven't quite figured out what direction I'm heading here -- just trying to figure out all the constraints placed on these types and not sure how I can get things lined up yet.